Pre commit pipeline

LukasMoll opened the pull request:
Added pre-commit pipeline config

#### Brief description of what is fixed or changed
Added a config file for the pre-commit pipeline.
This will run the linters when you commit.
![](https://ljvmiranda921.github.io/assets/png/tuts/precommit_pipeline.png)
Install pre-commit: `pip install pre-commit`
Install pre-commit in project: `pre-commit install`

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

Commit:
Added myself to .mailmap

oscarbenjamin commented:
There have been discussions about black before (gh-17287).

In any case what this pre-commit hook would do is reformat the whole of any file that a contributor edits. That means that any small change in one file turns into reformatting the whole file. We can't add something like that when most of the codebase is currently in a form that black would reformat.

Adding flake8 as a precommit hook seems fine though since that is already tested in CI.

LukasMoll commented:
Thanks for the feedback.
I removed black and updated the flake8 version.

oscarbenjamin commented on pyproject.toml:
```diff
@@ -70,7 +70,22 @@ exclude = [
 
 # Black default, although irrelevant with E501 ignored
+[tool.black]
 line-length = 88
```
This is still adding configuration for black and possibly changing the configuration of ruff.

LukasMoll replied:
I removed it.

brocksam commented on pyproject.toml:
```diff
@@ -70,7 +70,22 @@ exclude = [
 ]
 
 # Black default, although irrelevant with E501 ignored
+[tool.black]
 line-length = 88
+include = '\.pyi?$'
+exclude = '''
+/(
+    \.git
+  | \.hg
+  | \.mypy_cache
+  | \.tox
+  | \.venv
+  | _build
+  | buck-out
+  | build
+  | dist
+)/
+'''
```
This should be reverted with the removal of Black. It also (I think accidentally) changed the Ruff configuration - the `line-length = 88` is part of the Ruff configuration and so needs to remain under the `[tool.ruff]` heading.

LukasMoll replied:
Thanks for the clarification.
I reverted it.

brocksam reviewed (commented):
If we're having flake8 as a pre-commit hook then we should also have Ruff given that that's used to lint in CI too. Note that running flake8 on the whole SymPy repository takes 1-2 minutes.

oscarbenjamin commented:
> Note that running flake8 on the whole SymPy repository takes 1-2 minutes.

I think with this pre-commit hook it only runs flake8 on the files that are staged for commit (same would be for black). The check was quick when I ran it but running flake8 on the whole codebase is very slow on this computer (longer than 1-2 minutes).

Commit:
Added pre-commit pipeline config

This will run the linters inside the config when you commit
Install pre-commit: pip install pre-commit
Install pre-commit in project: pre-commit install

brocksam reviewed (commented):
I'd also like to see Ruff alongside flake8 in the `.pre-commit-config.yaml`. See [here](https://github.com/charliermarsh/ruff#usage) for how to configure for Ruff.

brocksam commented:
The `sdist-check` CI job is also failing because you've added a new file that isn't being accounted for in `release/compare_tar_against_git.py`. You'll need to add `.pre-commit-config.yaml` to the `git_whitelist` `set` in that file.

brocksam commented:
You also need to make sure that your release notes are correct. You've deleted half of the boilerplate in the original comment on this PR. You will need to edit it so that it reads:

\<!-- BEGIN RELEASE NOTES -->
NO ENTRY
\<!-- END RELEASE NOTES -->

(As this PR changes things that only relate to developers you can use "NO ENTRY")

Commit:
Added pre-commit config filename to the git_whitelist

brocksam commented on release/compare_tar_against_git.py:
```diff
@@ -55,6 +55,7 @@ def main(tarname, gitroot):
     'codecov.yml',
     'MANIFEST.in',
     'banner.svg',
+    '.pre-commit-config.yaml'
```
```suggestion
    '.pre-commit-config.yaml',
```

Skylion007 commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,9 @@
+repos:
+-   repo: https://github.com/PyCQA/flake8
+    rev: 6.0.0
```
You should add quotes around the rev for flake8 too

asmeurer replied:
Is pinning required? I don't think we pin on CI meaning this would get out of date.

LukasMoll replied:
I think it might interesting to have the pre-commit config and let the user decide which one they want to use. Pre-commits main advantage is it being quick blocker at the commit stage.

brocksam reviewed (commented):
Undoing my blocking review as this now looks good once the final comments are addressed.

asmeurer commented:
I think this is fine. It's optional and it only tests things that would be tested on CI anyway.

Commit:
Added myself (github) to .mailmap

Commit:
Added comma

Commit:
Added quotes

wermos commented:
> I think this is fine. It's optional and it only tests things that would be tested on CI anyway.
+1 from my side. I often forget to check code quality using `bin/test code_quality` and then have to make extra commits for fixing the code quality.

Skylion007 commented:
Can we enroll the repo in https://pre-commit.ci/ as well? I think the only thing we would need to do additionally is check the pre-commit CI parameters in the YAML and change any default ones that do not suit us. Or is that the current plan with this PR?

oscarbenjamin commented:
> Can we enroll the repo in https://pre-commit.ci/ as well?

What effect would that have?

Skylion007 commented:
> > Can we enroll the repo in https://pre-commit.ci/ as well?
> 
> What effect would that have?

Would create a bot that runs pre-commit on every PR and would automatically commit fixes back to the PR.

oscarbenjamin commented:
> Would create a bot that runs pre-commit on every PR and would automatically commit fixes back to the PR.

What kind of fixes? This seems like something that could be useful but also controversial. We need to be clear about exactly what this bot would do.

Also who does it commit as?

oscarbenjamin commented:
> > Also who does it commit as?
> 
> Pre-commit never commits changes.

I was referring to `pre-commit.ci`:
https://pre-commit.ci/

You can see that it commits fixes if configured to do so. We could have pre-commit run `bin/strip_whitespace` and maybe some ruff autofixes like F401 "unnecessary import" (a common cause of CI failure).

Another possibility is automatically reordering the .mailmap file. Actually if `pre-commit` could stop someone from making commits with the wrong name and email in the first place then that would help a lot of new contributors who otherwise struggle with the mailmap check.

> I think it's nice to have because we can remove some code from our CI config files and we ensure that the pre-commit and CI are the same.

Agreed. Although if we expect anyone to use the `pre-commit` git hook as an actual git hook then we need to make sure it stays fast.

Skylion007 commented:
@oscarbenjamin Better yet, pre-commit has native hooks to strip whitespace and other useful common formatter actions. See this repo for some: https://github.com/pre-commit/pre-commit-hooks

oscarbenjamin commented:
We discussed pre-commit.ci on the mailing list and the feeling seems to be that we would not want to use that:
https://groups.google.com/g/sympy/c/RcgxZxbEu6c/m/2BQcszZpAgAJ

However using pre-commit to run the quality checks in CI could work.

Maybe we could change the CI config to use `pre-commit run -a` instead of running flake8 and ruff directly.

oscarbenjamin commented:
> Maybe we could change the CI config to use `pre-commit run -a`

Actually I just tried that and it failed:
```python
$ pre-commit run -a
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

.ci/parse_durations_log.py:18:13: E722 do not use bare 'except'
bin/ask_update.py:19:1: F401 'pprint' imported but unused
bin/mailmap_check.py:17:1: F401 'os' imported but unused
...
```
I'm not sure what the appropriate command is to have pre-commit check all files in say the sympy directory.

Skylion007 commented:
> > Maybe we could change the CI config to use `pre-commit run -a`
> 
> Actually I just tried that and it failed:
> 
> ```python
> $ pre-commit run -a
> flake8...................................................................Failed
> - hook id: flake8
> - exit code: 1
> 
> .ci/parse_durations_log.py:18:13: E722 do not use bare 'except'
> bin/ask_update.py:19:1: F401 'pprint' imported but unused
> bin/mailmap_check.py:17:1: F401 'os' imported but unused
> ...
> ```
> 
> I'm not sure what the appropriate command is to have pre-commit check all files in say the sympy directory.

Oh this is easy, you need to either include those files in the flake8 config or set `pass_filenames: false` to the pre-commit hook config for that hook. Pre-commit will pass on the files that have changed to flake8, which is useful but will override the flake8 config for which files to check.

Skylion007 commented:
> > Setting it to false will have flake8 run normally, (ie. just calling`flake8`), but then it will check all the files even the unchanged ones.
> 
> We still want `pre-commit` to check changed files only when run locally. Just in CI we want it to check everything.

Then you need to set either the `exclude` or `files` keys properly on the pre-commit hook.

asmeurer commented:
Why don't we just fix the flake errors in these Python files that are outside of `sympy/`? This should be a very straightforward thing to fix.

oscarbenjamin commented:
Yes, it should be straight-forward to fix the lint errors outside of the sympy directory. It would be much better to do that than complicate the config by trying to avoid avoid those files.

Commit:
Added pre-commit to the github CI + fixed pre-commit errors

LukasMoll commented:
I added some things to the pre-commit and replaced it with the current CI implementation to keep it all the same.
Fixed the pre-commit errors.

oscarbenjamin commented on .github/workflows/runtests.yml:
```diff
@@ -23,16 +23,13 @@ jobs:
-      - name: Run Ruff on the sympy package
-        run: ruff check .
+      - name: Run flake8 and ruff on the sympy package with pre-commit
+        run: pre-commit install; pre-commit run -a
```
Is the install step actually needed here?

LukasMoll replied:
Yes, you need to install pre-commit in the current repository before you can run it.

oscarbenjamin replied:
Are you sure about that?

To be clear if this change is merged I would like to be able to run pre-commit but I would not want to install the git hook.

LukasMoll replied:
Sorry, I think you're right. You can still run it when it's not installed.

oscarbenjamin commented on .github/workflows/runtests.yml:
```diff
@@ -23,16 +23,13 @@ jobs:
       - run: python -m pip install --upgrade pip
 
-      - run: pip install mpmath flake8 flake8-comprehensions ruff pytest
+      - run: pip install mpmath pytest pre-commit
```
Is pytest needed here?

oscarbenjamin replied:
Oh, maybe it's needed for `bin/test quality`.

Ideally we would fold those checks into `pre-commit` as well in future.

LukasMoll replied:
That's a good idea. I'll implement it. I know that some do not work/are to slow (e.g. mypy) so it might not be possible for everything.

oscarbenjamin commented:
Is it possible to use a requirements.txt or constraints.txt file rather than pinning the versions in the pre-commit config. I'm just thinking that it might be easier to keep that up to date along with other requirements files using pip-upgrader.

oscarbenjamin commented:
I think the antlr files are autogenerated and should be left unchanged.

LukasMoll commented:
> Is it possible to use a requirements.txt or constraints.txt file rather than pinning the versions in the pre-commit config. I'm just thinking that it might be easier to keep that up to date along with other requirements files using pip-upgrader.

Good idea. I'm not 100% sure but this should be possible.

Commit:
removed unecessary pre-commit install

Commit:
Added local linters to pre-commit

Commit:
Removed unecessary checks that were moved to pre-commit

LukasMoll commented:
[It seems to be not possible to pin the version in a file like requirements.txt.](https://stackoverflow.com/questions/65940222/pre-commit-com-same-version-in-pre-commit-config-yaml-and-requirements-txt)

oscarbenjamin commented:
> It seems to be not possible to pin the version in a file like requirements.txt.

As an alternative is is possible to use the precommit.ci bot to update versions in the precommit config but without using its "autofix" feature?

LukasMoll committed:
Merge branch 'master' into pre-commit-pipeline
